api/server/errorhandler.go

```diff
@@ -17,18 +22,99 @@ func makeErrorHandler(err error) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
 		statusCode := httpstatus.FromError(err)
 		vars := mux.Vars(r)
-		if apiVersionSupportsJSONErrors(vars["version"]) {
-			response := &types.ErrorResponse{
-				Message: err.Error(),
-			}
-			_ = httputils.WriteJSON(w, statusCode, response)
-		} else {
+		version := vars["version"]
+
+		const firstAPIVersionWithJSONErrors = "1.24"
+		if version != "" && versions.LessThan(version, firstAPIVersionWithJSONErrors) {
 			http.Error(w, status.Convert(err).Message(), statusCode)
+			return
+		}
+
+		response := marshalErrorResponse(err)
+		const firstAPIVersionWithJSONMultiErrors = "1.44"
+		if version != "" && versions.LessThan(version, firstAPIVersionWithJSONMultiErrors) {
+			response.Errors = nil
+		}
+
+		_ = httputils.WriteJSON(w, statusCode, response)
+	}
+}
+
+type joiningError interface {
+	error
+	Unwrap() []error
+}
+
+type wrappingError interface {
+	error
+	Unwrap() error
+}
+
+// marshalErrorResponse returns a tree of [*types.ErrorResponse] representing the argument err.
+func marshalErrorResponse(err error) *types.ErrorResponse {
+	if errdefs.IsHTTPOnlyError(err) {
```

Comment:
> When one HTTPOnlyError wraps another HTTPOnlyError, why are they marshaled differently? The inner one is marshaled as a wrappingError while the outer is discarded.

Hum, nope they would be both discarded and only the inner non-HTTPOnlyError would be marshaled.

The initial intent was to discard "HTTPOnlyError" as they don't add any context. If I remove this code, here's what it produces:

```golang
// Currently
err := errdefs.InvalidParameter(errors.New("foobar"))
expected := &types.ErrorResponse{Message: "foobar"}

err := errdefs.InvalidParameter(errdefs.InvalidParameter(errors.New("foobar")))
expected := &types.ErrorResponse{Message: "foobar"}

// With the IsHTTPOnlyError check removed
err := errdefs.InvalidParameter(errors.New("foobar"))
expected := &types.ErrorResponse{Message: "foobar", Errors: []*types.ErrorResponse{{Message: "foobar"}}}

err := errdefs.InvalidParameter(errdefs.InvalidParameter(errors.New("foobar")))
expected := &types.ErrorResponse{Message: "foobar", Errors: []*types.ErrorResponse{{Message: "foobar"}}}
```

Do we want to keep this child `ErrorResponse` although it doesn't add value?

Reply:
Now try marshaling `errdefs.InvalidParameter(errdefs.Conflict(errors.New("foobar")))`. https://github.com/moby/moby/blob/63d477b20e77fdf69a50701315c17a84733d1775/errdefs/helpers.go#L38-L41

> Do we want to keep this child `ErrorResponse` although it doesn't add value?

I wouldn't want to keep HTTPOnly errors in the marshaled error tree. I just want them to be marshaled (or discarded) consistently, independent of their position in the error tree.

Reply:
Ah, indeed. I see your point. So I should probably recursively unwrap HTTP-only errors here.

```golang
	if errdefs.IsHTTPOnlyError(err) {
		if wrappedErr := errors.Unwrap(err); wrappedErr != nil {
			return marshalErrorResponse(wrappedErr)
		}
	}
```

Reply:
I'm a bit confused by this part; so currently, this PR adds no additional information about the unwrapped errors (other than splitting the nested error messages). but wouldn't the type of error produced be _actually_ useful once we add more details?

I don't think we should consider these to be "HTTP-only" error messages, because they _aren't_. Mapping them to a HTTP status code is _one_ purpose, but that's a translation of the "kind" of error returned (e.g. an object was not found).

If we _are_ considering this response to be used for handling specific conditions, wouldn't that information be exactly what's useful? (e.g., we returned an invalid request error, because an object was not found).

Reply:
To some extend I'm wondering; we added these nested errors because Go's original formatting (omitting the "2 errors") was not ideal, so we added the option to unwrap those errors, so that they could be formatted in different ways, which a fallback for clients that don't support that (in the form of the existing pre-formatted `Message`.

- But currently, there's no such implementation on the client side (should this PR include that part?).
- If we anticipate this structured format to be used for more in-depth error analysing, I'm wondering if a stack-trace for errors would be more useful 🤔 I know I gave that a very quick attempt in https://github.com/moby/moby/pull/43252

Reply:
The client implementation can be added in a follow-up PR. There's no need to increase the scope of this one. The API change is backwards-compatible, after all.

Stack traces are not either/or. A stack trace is a property of an individual error. This PR adds support for marshaling a list/tree of errors. They are orthogonal concerns.

Reply:
As far as the API is concerned, mapping to an HTTP status code _is_ their only purpose. [The client can losslessly recover that information](https://github.com/moby/moby/blob/2b449e0e6512c70ac783450ffb891d940c29de51/errdefs/http_helpers.go#L8) from the HTTP status code the error is mapped to, combined with the `Message` of the unwrapped error.

Reply:
> The client implementation can be added in a follow-up PR. There's no need to increase the scope of this one. The API change is backwards-compatible, after all.

So what's currently the use for these structured errors if there's nothing consuming them? I know this work started because the default formatting of Golang's native "multi-errors" was not great (missing the `2 errors occurred` part. And that's addressed in this PR by using our own formatting (through the `formatErrors()` function)

Do we _need_ the other parts already, or should we split the `formatErrors()` part to a separate PR, and draft a more in-depth plan what we will need the structured _API response_ for?